### PR TITLE
[Text Generation] KVCacheStorage Implementation

### PR DESCRIPTION
**Pull request by dbogunowicz**
Todo:
- [ ]  Get basic tests in place before we land this PR

Quick manual testing:

```python
from deepsparse import Pipeline

def _test_pipeline(engine_type):
    opt = Pipeline.create(task="opt",
                          model_path="/home/ubuntu/damian/sparseml/deployment",
                          max_generated_tokens = 10,
                          engine_type = engine_type,
                          use_deepsparse_cache = False)
    print('----------')
    prompt = "My name is Richard Dickson and I am a data scientist"
    out = opt(sequences=prompt, session_ids = "richard", return_logits = True)
    print(prompt + out.sequences[0])
    print('---------')
    prompt = "As I mentioned, my name is Richard"
    out = opt(sequences=prompt,session_ids = "richard", return_logits=True)
    print(prompt + out.sequences[0])
    print('---------')
    prompt = "And my job is"
    out = opt(sequences=prompt,session_ids = "richard", return_logits=True)
    print(prompt + out.sequences[0])

_test_pipeline(engine_type ="onnxruntime")
```

```bash
----------
My name is Richard Dickson and I am a data scientist. I am a software engineer and I am a
---------
As I mentioned, my name is Richard Dickson.



I am a
---------
And my job is a data scientist and I am a data scientist and
```

### CODE REVIEW

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 1]
```diff
@@ -93,27 +93,12 @@ def __init__(
...
         self.sampling_temperature = sampling_temperature
         self.deterministic = deterministic
+        self.use_deepsparse_cache = use_deepsparse_cache
         self.input_ids_length = input_ids_length
         self.kv_cache_enabled = kv_cache_enabled
-        self.kv_cache = (
-            DecoderKVCache(use_deepsparse_cache) if kv_cache_enabled else None
-        )
+        self.kv_cache_storage = SessionStorageKVCache() if kv_cache_enabled else None
+        self.session = None
```
Comment (bfineran) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:100]: session cannot be a property of the pipeline - it must exist only in the context of an inference after being grabbed from storage. if we're reading and writing sessions from the object then we run into race conditions when multiple inferences are run at the same time

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -55,11 +56,11 @@ class TextGenerationInput(BaseModel):
         "the logits for the input text sequence and the "
         "generated text sequence. ",
     )
-    session_id: Optional[str] = Field(
+    session_ids: Union[None, List[str], str] = Field(
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:59]: let's talk more on this - if we have a batched input does this mean each batch item should have its own session / kv cache stored?

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -72,6 +73,50 @@ class TextGenerationInput(BaseModel):
         "can compute metric in a batched fashion. ",
     )
 
+    @validator("session_ids")
+    def validate_session_ids(cls, value, values) -> Union[None, List[str]]:
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:77]: let's move this to a helper function in a separate file

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 4]
```diff
@@ -209,6 +254,8 @@ def __init__(
                 "Set `max_generated_tokens` to 1 to support that scenario."
             )
 
+        self._use_session_ids = False
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:257]: can you explain more what this is used/defaulted for?

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 5]
```diff
@@ -285,10 +332,12 @@ def process_inputs(self, inputs: TextGenerationInput) -> List[numpy.ndarray]:
         onnx_input_names = self.multitoken_engine.onnx_input_names_no_cache
         engine_input = self.tokens_to_engine_input(input_tokens, onnx_input_names)
 
-        if inputs.session_id is not None:
-            # if session_id is provided, we need to set it in engines
-            self.engine.session_id = inputs.session_id
-            self.multitoken_engine.session_id = inputs.session_id
+        self._use_session_ids = False
+        if inputs.session_ids is not None:
+            # TODO: This is ugly, we should find a better way to pass
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:337]: why not have `NLDecoderEngine` accept an additional kwarg for this? we will only be passing this in when the engine is an `NLDecoderEngine` right?

**Review by bfineran — CHANGES_REQUESTED** [review]
looks like a great start - let's set up time to go through this a bit more together

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 5 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:337]: yeah, outsourcing all the `session_id` management to `NLDecoderEngine` that early is a good idea.

File: src/deepsparse/transformers/engines/nl_decoder_engine.py (hunk 1 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/engines/nl_decoder_engine.py:100]: Please read carefully, this is not the pipeline attribute, this is the attribute of a `NLDecoderEngine`.

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 2 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:59]: yes, at least for now. happy to discuss the UI if this is not the case.

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 3 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:77]: yeah, good idea.

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 4 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:257]: To handle the session_ids management in the pipeline context. But given that we want to completely outsource session_id management to the `NLDecoderEngine`, this is probably not needed!

**Comment by dbogunowicz** [comment]
For anyone who may possibly run look into this PR when I am OOO: rebase `0524291` may not be clean.

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 6]
```diff
@@ -292,34 +311,38 @@ def update_kv_cache(
...
         inside the engine. We only update the number of tokens processed.
 
         :param kv_cache_state: The state of the kv cache storage
         :param input_ids_len: The length of input_ids
+        :param session_id: The session id to fetch the kv cache session for
         """
         if self.internal_cache_active:
-            self.kv_cache.total_num_processed_tokens += input_ids_len
+            session = self.kv_cache_storage.get(session_id)
+            # TODO: Check whether this gets updated correctly
```
Comment (bfineran) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:330]: let's make sure to resolve before merge

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 7]
```diff
@@ -406,14 +422,29 @@ def engine_forward(
         with self.timer_manager.new_timer_context(total_inference=False) as timer:
             streamer = context.get("streamer")
 
+            # pop the session id from the inputs
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:425]: let's add a clarifying comment here

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 8]
```diff
@@ -510,23 +560,29 @@ def prompt_inference(
...
             (with dimensions ['batch_size', 'num_tokens', 'vocab_size'])
         """
+        num_cached_entries = self.engine.num_non_blank_cache_entries(session_id)
 
         new_token = tokens[-1]
         # padding is added to left, so attention mask is 1s from the
         # right up to the number of total tokens (prompt + generated)
         attention_mask = numpy.zeros((1, self.sequence_length), dtype=numpy.int64)
-        num_tokens_processed = min(len(tokens), self.sequence_length)  # cap by seq len
+        # TODO: We probably don't need this line below but let's see
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:580]: resolve before landing

File: src/deepsparse/transformers/utils/decoder_kv_cache.py [hunk 9]
```diff
@@ -241,7 +238,11 @@ def num_non_blank_entries(self):
         """
         :return: the number of non-blank entries in the kv cache
         """
-        return min(self.capacity, self.total_num_processed_tokens)
+
+        # given a 2D array get number of rows that are not entirely zero
+        single_kv_cache_entry = self.cached_inputs[list(self.cached_inputs.keys())[0]]
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/decoder_kv_cache.py:243]: why are we calculating this indirectly instead of just tracking the number of total tokens for the session as an attribute

Comment (bfineran) [reply src/deepsparse/transformers/utils/decoder_kv_cache.py:243]: additionally - let's make sure calculating positions correctly - it shouldn't have anything with blank entries or padding, just the number of real tokens

File: src/deepsparse/transformers/utils/storage_kv_cache.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+class SessionStorageKVCache:
+    """
+    A storage that stores the kv cache sessions.
+    Each session is a DecoderKVCache object that stores the state of the kv cache.
+    The storage is a set of all the active sessions.
+    """
+
+    def __init__(self):
+        self._memory: Set[DecoderKVCache] = set()
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/storage_kv_cache.py:35]: let's use dictionary, no reason we can't key by an attribute

**Review by bfineran — CHANGES_REQUESTED** [review]

**Review by bfineran — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 11]
```diff
@@ -448,14 +465,40 @@ def engine_forward(
         with self.timer_manager.new_timer_context(total_inference=False) as timer:
             streamer = context.get("streamer")
 
+            # engine_inputs is a list of numpy arrays plus additional
+            # session_id string. We need to pop the session_id string
+            # and from the engine_inputs. The session_id will be used
+            # seperately to keep track of the appropriate kv cache session
+            # (if kv cache is enabled)
+            session_id = engine_inputs.pop(
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:473]: Sara: what if there is another string that we are adding. 
Better: engine_inputs.pop(-1)

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 12]
```diff
@@ -167,15 +163,17 @@ def run(self, inputs: List[numpy.ndarray], val_inp: bool) -> List[numpy.ndarray]
             if val_inp:
                 self.engine._validate_inputs(inputs)
             # run the engine with the LIB.kv_cache object
-            return self.engine._eng_net.execute_list_out(
-                inputs, self.kv_cache.engine_internal_cache
-            )
+            engine_internal_cache = self.kv_cache_storage.get(
+                session_id
+            ).engine_internal_cache
+            return self.engine._eng_net.execute_list_out(inputs, engine_internal_cache)
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:169]: potentially puting a lock on session storage...

**Review by Satrat — DISMISSED** [review]

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 13]
```diff
@@ -173,15 +170,19 @@ def run(self, inputs: List[numpy.ndarray], val_inp: bool) -> List[numpy.ndarray]
             # we want to pass the empty kv cache inputs
             # (batch_size=0) to the engine. Therefore,
             # we skip the validation
-            return self.engine._eng_net.execute_list_out(
-                inputs, self.kv_cache.engine_internal_cache
-            )
+
+            # run the engine with the LIB.kv_cache object
+            engine_internal_cache = self.kv_cache_storage.get(
```
Comment (bfineran) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:175]: Can we guarantee this id exists?

File: src/deepsparse/transformers/utils/helpers.py [hunk 14]
```diff
@@ -56,6 +57,44 @@ def repeat_inputs(
     return repeated_seq
 
 
+def validate_session_ids(
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/helpers.py:60]: Let's go over how edge cases are handled in our meeting

File: src/deepsparse/transformers/utils/storage_kv_cache.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    def __len__(self):
+        return len(self._memory)
+
+    def __str__(self):
+        return (
+            f"{SessionStorageKVCache.__name__}:\n "
+            f"\tsessions: {[session_name for session_name in self._memory.keys()]}\n"
+        )
+
+    def has_session(self, session_id: str) -> bool:
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/storage_kv_cache.py:47]: Could define as __contains__

File: src/deepsparse/transformers/utils/storage_kv_cache.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+        :param session_id: The identifier of the cache session.
+        :return: The state of the kv cache for the session.
+        """
+        session = self._memory.get(session_id)
+        if session is None:
+            _LOGGER.debug(f"No cache session found for session id: {session_id}")
+        return session
+
+    def pop(self, session_id: str):
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/storage_kv_cache.py:81]: Pop should return the removed value

**Review by Satrat — APPROVED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 17]
```diff
@@ -749,29 +810,32 @@ def engine_inputs_for_prefill(
...
+
                 elif name == "causal_mask":
                     # delay creation of the causal mask
                     continue
                 elif name == "positions":
-                    engine_input = (
-                        numpy.arange(
-                            num_cached_entries,
-                            num_cached_entries + self.prompt_sequence_length,
+                    if self.prompt_sequence_length == 1:
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:823]: Remove this!

**Comment by dbogunowicz** [comment]
Outdated, closing.
